coderd/database/queries/insights.sql

```diff
@@ -771,3 +771,124 @@ SELECT
 FROM unique_template_params utp
 JOIN workspace_build_parameters wbp ON (utp.workspace_build_ids @> ARRAY[wbp.workspace_build_id] AND utp.name = wbp.name)
 GROUP BY utp.num, utp.template_ids, utp.name, utp.type, utp.display_name, utp.description, utp.options, wbp.value;
+
+-- name: GetUserStatusCountsOverTime :many
+-- GetUserStatusCountsOverTime returns the count of users in each status over time.
+-- The time range is inclusively defined by the start_time and end_time parameters.
+--
+-- Bucketing:
+-- Between the start_time and end_time, we include each timestamp where a user's status changed or they were deleted.
+-- We do not bucket these results by day or some other time unit. This is because such bucketing would hide potentially
+-- important patterns. If a user was active for 23 hours and 59 minutes, and then suspended, a daily bucket would hide this.
+-- A daily bucket would also have required us to carefully manage the timezone of the bucket based on the timezone of the user.
+--
+-- Accumulation:
+-- We do not start counting from 0 at the start_time. We check the last status change before the start_time for each user. As such,
+-- the result shows the total number of users in each status on any particular day.
+WITH
+-- dates_of_interest defines all points in time that are relevant to the query.
+-- It includes the start_time, all status changes, all deletions, and the end_time.
+dates_of_interest AS (
+	SELECT @start_time::timestamptz AS date
+
+	UNION
+
+    SELECT DISTINCT changed_at AS date
+    FROM user_status_changes
+    WHERE changed_at > @start_time::timestamptz
+        AND changed_at < @end_time::timestamptz
```

Comment:
If it is inclusive, should it be?
```suggestion
    WHERE changed_at >= @start_time::timestamptz
        AND changed_at <= @end_time::timestamptz
```

Reply:
> I think that's covered by the initial and final union.

Correct. 

> Monday 00:00 -> Monday 00:00 instead of subtracting an arbitrary unit of time or using an end-of-day function

I do not think this graph is such a critical feature that we need to split hairs over 1ns at the end of the range. There are higher impact changes to make. If you'd like me to change it, please ask me directly to do so and I will. My own opinion is that for how this chart is going to be displayed inclusive works well enough.